modules/indexer/codes/indexer.go

```diff
@@ -0,0 +1,383 @@
+	if err != nil {
+		return err
+	} else if !notEmpty {
+		return nil
+	}
+
+	// if there is any existing repo indexer metadata in the DB, delete it
+	// since we are starting afresh. Also, xorm requires deletes to have a
+	// condition, and we want to delete everything, thus 1=1.
+	if err := models.DeleteAllRecords("repo_indexer_status"); err != nil {
+		return err
+	}
+
+	maxRepoID, err := models.GetMaxID("repository")
+	if err != nil {
+		return err
+	}
+
+	log.Info("Populating the repo indexer with existing repositories")
+	// start with the maximum existing repo ID and work backwards, so that we
+	// don't include repos that are created after gitea starts; such repos will
+	// already be added to the indexer, and we don't need to add them again.
+	for maxRepoID > 0 {
+		repos := make([]*models.Repository, 0, models.RepositoryListDefaultPageSize)
+		err = models.FindByMaxID(maxRepoID, models.RepositoryListDefaultPageSize, &repos)
+		if err != nil {
+			return err
+		} else if len(repos) == 0 {
+			break
+		}
+		for _, repo := range repos {
+			codesIndexerQueue.Push(&IndexerData{
+				RepoID:   repo.ID,
+				Filepath: repo.RepoPath(),
+				IsDelete: false,
+			})
+			maxRepoID = repo.ID - 1
+		}
+	}
+	log.Info("Done populating the repo indexer with existing repositories")
+	return nil
+}
+
+// repoChanges changes (file additions/updates/removals) to a repo
+type repoChanges struct {
+	Updates          []fileUpdate
+	RemovedFilenames []string
+}
+
+type fileUpdate struct {
+	Filename string
+	BlobSha  string
+}
+
+func getDefaultBranchSha(repo *models.Repository) (string, error) {
+	stdout, err := git.NewCommand("show-ref", "-s", repo.DefaultBranch).RunInDir(repo.RepoPath())
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(stdout), nil
+}
+
+// getRepoChanges returns changes to repo since last indexer update
+func getRepoChanges(repo *models.Repository, revision string) (*repoChanges, error) {
+	if err := repo.GetIndexerStatus(); err != nil {
+		return nil, err
+	}
+
+	if len(repo.IndexerStatus.CommitSha) == 0 {
+		return genesisChanges(repo, revision)
+	}
+	return nonGenesisChanges(repo, revision)
+}
+
+func addUpdate(update fileUpdate, repo *models.Repository, batch rupture.FlushingBatch) error {
+	stdout, err := git.NewCommand("cat-file", "-s", update.BlobSha).
+		RunInDir(repo.RepoPath())
+	if err != nil {
+		return err
+	}
+	if size, err := strconv.Atoi(strings.TrimSpace(stdout)); err != nil {
+		return fmt.Errorf("Misformatted git cat-file output: %v", err)
+	} else if int64(size) > setting.Indexer.MaxIndexerFileSize {
+		return nil
+	}
+
+	fileContents, err := git.NewCommand("cat-file", "blob", update.BlobSha).
+		RunInDirBytes(repo.RepoPath())
+	if err != nil {
+		return err
+	} else if !base.IsTextFile(fileContents) {
+		return nil
+	}
+	return batch.Index(
+		filenameIndexerID(repo.ID, update.Filename),
+		&IndexerData{
+			RepoID:   repo.ID,
+			Filepath: update.Filename,
+			Content:  string(fileContents),
```

Comment:
You may want to make sure the file contents is valid utf-8. See #7809 and #7814.